api/classes/class-lightweight-api.php

```diff
@@ -227,161 +226,580 @@ public function delete_transient( $name ) {
+			}
+
+			foreach ( $new_views as $read_event ) {
+				if ( ! isset( $reader_data['views'][ $read_event['context'] ] ) ) {
+					$reader_data['views'][ $read_event['context'] ] = 0;
+				}
+				$reader_data['views'][ $read_event['context'] ] ++;
+
+				if ( isset( $read_event['value']['categories'] ) ) {
+					$categories = explode( ',', $read_event['value']['categories'] );
+					if ( ! isset( $reader_data['category'] ) ) {
+						$reader_data['category'] = [];
+					}
+					foreach ( $categories as $term_id ) {
+						if ( ! isset( $reader_data['category'][ $term_id ] ) ) {
+							$reader_data['category'][ $term_id ] = 0;
+						}
+						$reader_data['category'][ $term_id ] ++;
+					}
+				}
+			}
+
+			$this->save_reader( $client_id, $reader_data );
+		}
+
+		// Ensure client ID exists and keys match across all new items.
+		$reader_events_blueprint = $this->reader_events_blueprint;
+		$events                  = array_map(
+			function( $item ) use ( $client_id, $reader_events_blueprint ) {
+				$item['client_id'] = $client_id;
+
+				if ( empty( $item['date_created'] ) ) {
+					$item['date_created'] = gmdate( 'Y-m-d H:i:s' );
+				}
+
+				return wp_parse_args( $item, $this->reader_events_blueprint );
+			},
+			$events
+		);
+
+		// If ignoring cache, write directly to DB.
+		if ( $this->ignore_cache ) {
+			$write_result = $this->bulk_db_insert( $events );
+
+			// If DB write was successful, rebuild cache.
+			if ( $write_result ) {
+				$this->debug['write_query_count'] += 1;
+			} else {
+				return false;
+			}
+		} else {
+			// Write items to the flat file to be parsed to the DB at a later point.
+			Segmentation_Report::log_reader_events( $events );
+		}
+
+		// Rebuild cache.
+		$all_events = array_merge( $existing_events, $events );
+
+		return wp_cache_set(
+			'reader_events',
+			$all_events,
+			$client_id
+		);
 	}
 
 	/**
-	 * Retrieve all clients' data.
+	 * Retrieve a large sample of readers' data.
 	 *
-	 * @return array All clients' data.
+	 * @return array Readers' data.
 	 */
-	public function get_all_clients_data() {
+	public function get_all_readers_data() {
 		global $wpdb;
-		$events_table_name = Segmentation::get_events_table_name();
+		$readers_table_name = Segmentation::get_readers_table_name();
 
 		// Results are limited to the 1000 most recent rows for performance reasons.
-		$all_client_ids_rows = $wpdb->get_results( "SELECT DISTINCT client_id,id FROM $events_table_name ORDER BY id DESC LIMIT 1000" ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
+		$all_client_ids_rows = $wpdb->get_results( "SELECT DISTINCT client_id,date_modified FROM $readers_table_name WHERE is_preview IS NULL ORDER BY date_modified DESC LIMIT 1000" ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching, WordPress.DB.PreparedSQL.InterpolatedNotPrepared
```

Comment:
Are we ok not retrieving all data when the total exceeds 1000?

Reply:
This function is only used to calculate approximate reach/viewability of a segment. The query is limited to 1000 to avoid performance issues on very large sites that will have many thousands of readers.